### PR TITLE
Recipe: Dependency loop detection

### PR DESCRIPTION
**Pull request by Lotes**
I added a recipe for detecting any kind of dependency loop and resolving acyclic dependencies.

It is not specific to Langium, but it is one of these problems, that might need to be solved for several language features. Like recursion detection or recursive data structures.

The detection is implemented as a Langium validation. I think this is useful for users, that have never came across with such graph algorithms.

### CODE REVIEW

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+checkGreetingCycles(model: Model, accept: ValidationAcceptor): void {
+    //arrange the graph
+    const graph = new DirectedGraph<{}, {greeting: Greeting}>();
+    model.persons.forEach(person => {
+        graph.addNode(person.name);
+    })
+    model.greetings.forEach(greeting => {
+        if(greeting.greeter.ref && greeting.greeted.ref && !graph.hasDirectedEdge(greeting.greeter.ref.name, greeting.greeted.ref.name)) {
+            graph.addEdge(greeting.greeter.ref.name, greeting.greeted.ref.name, {
+                greeting
```
Comment (JohannesMeierSE) [inline hugo/content/docs/recipes/validation/dependency-loops.md:102]: maybe add a comment in the code, that we need to store the AST node in order to attach a warning to it later?

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+    });
+    
+    //compute the components
+    const components = stronglyConnectedComponents(graph);
+
+    //evaluate result (filter out size-1-components)
+    const actualLoops = components.filter(c => c.length > 1);
+    for (const component of actualLoops) {
+        const set = new Set<string>(component);
+        //for each node in the component...
```
Comment (JohannesMeierSE) [inline hugo/content/docs/recipes/validation/dependency-loops.md:114]: // for each node (not edge!) in the component...

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 3]
```diff
@@ -0,0 +1,258 @@
...
+        for (const from of set) {
+            //check whether the out edges...
+            for (const { target: to, attributes: { greeting } } of graph.outEdgeEntries(from)) {
```
Comment (JohannesMeierSE) [inline hugo/content/docs/recipes/validation/dependency-loops.md:117]: I am not sure about the following two cases:
- What about the case, that there are two same greetings "Homer greets Marge! Homer greets Marge!" Will both be marked?
- Is each edge marked exactly once (and not twice) for the example "A greets B! A greets C! B greets D! C greets D!"?

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+
+#### Complex detection
+
+If you have a `n:m` relationship like it is given for function calls (a function can be called by `m` function and can call `n` functions), you can solve the question for loops by creating a directed graph.
+
+In the example of function calls you have to store edges for each call from function A to every function B. The nodes are the set of all functions. The key algorithm is the search for the so-called strongly-connected components in the resulting graph.
+
+It is recommended not to implement your own version of that algorithm. Please use an existing solution! The algorithm is able to output every loop with all its members of that loop. Perfect for adding a custom validation!
+
+#### Complex resolution
```
Comment (JohannesMeierSE) [inline hugo/content/docs/recipes/validation/dependency-loops.md:41]: During the first reading, I didn't got, that this is a follow-up step after ensuring, that there are no cycles: Maybe, it was only me, maybe, there is some motivation missing at the beginning?

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 5]
```diff
@@ -0,0 +1,258 @@
+---
+title: "Dependency loops"
+weight: 100
+---
+
+## What is the problem?
```
Comment (JohannesMeierSE) [inline hugo/content/docs/recipes/validation/dependency-loops.md:6]: My personal impression is, that the motivation is not clear enough. I write some ideas in order to clarify this impression:

What about supporting some examples with more details, e.g. the Langium grammar for dependencies

```langium
Component: name=ID ('depends on' dependencies+=[Component]*)?;
```

or a class diagram for the resulting TS structure

```mermaid
classDiagram
class Component {
dependencies: Component[*]
}
```

?

The purpose, i.e. to create errors for the users, is missing.

List some more examples and name this examples more concrete:
- no cyclic data flow in flow charts
- no cyclic imports of dependencies
- no cycles in inheritance hierarchies of data types/types/classes
- no cycles in function calls (might not be supported by some architectures)

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 6]
```diff
@@ -0,0 +1,258 @@
...
+
+If you are building composite data structures or have some computation flow, you might be interested whether the product, that you are generating, does contain any loops back to the already used product.
+
+For data structures you could think of a structure in C that references itself (without using the pointer notation). This would lead to an infinitely expanding data type, which is practically not doable.
+
+Or for control flows these loops can be interpreted as recursion detection, a function that calls itself.
```
Comment (JohannesMeierSE) [inline hugo/content/docs/recipes/validation/dependency-loops.md:12]: I didn't got this example, sorry

**Review by JohannesMeierSE — COMMENTED** [review]
Thank you very much for this recipe @Lotes!

It makes a lot of sense to have this recipe and I like the nice application for the hello-world example!

However, the first section with the motivation is not clear enough from my point of view. After reading the first sentences, it is not very clear, what the actual problem is.

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 7]
```diff
@@ -0,0 +1,258 @@
...
+
+* `Homer` is not greeted by anyone, so he can start greeting `Marge`.
+* `Marge` and `Pinky` are blocked by `Pinky` and `Brain`.
+* `Brain` is the next and unblocks `Pinky`.
+* After `Pinky` is done, `Marge` is unblocked as well.
+* But `Marge` has no one to greet.
+* So, we are done.
+
+## Appendix
+
```
[the remark this answers is not in the document]
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md:180]: I think repetition is better than omitted information. You have to actively search for the missing pieces again.

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 8]
```diff
@@ -0,0 +1,258 @@
...
+### Complex nature
+
+#### Complex detection
+
+If you have a `n:m` relationship like it is given for function calls (a function can be called by `m` function and can call `n` functions), you can solve the question for loops by creating a directed graph.
+
+In the example of function calls you have to store edges for each call from function A to every function B. The nodes are the set of all functions. The key algorithm is the search for the so-called strongly-connected components in the resulting graph.
+
+It is recommended not to implement your own version of that algorithm. Please use an existing solution! The algorithm is able to output every loop with all its members of that loop. Perfect for adding a custom validation!
```
[the remark this answers is not in the document]
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md:39]: I try to explain the algorithmic solution, which is independent from the validation.

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 9]
```diff
@@ -0,0 +1,258 @@
...
+
+#### Complex resolution
+
+The directed graph approach can be processed further when there were no loops found:
```
[the remark this answers is not in the document]
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md:46]: A bidirectional graphs without cycles are just trees. I don't know how we should reflect this. The recipe is about dependencies that by definition have a direction: s.o. who is dependent and s.o. who this guy depends on.

A friendship graph could be such a thing. You would have a cycle when at least one person knows two other persons that know each other as well.... wdyt? Any better example?

File: hugo/content/docs/recipes/validation/dependency-loops.md (hunk 3 above)
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md:124]: That depends on the graph implementation. Graphology has different implementations for "Single"/MultiGraphs.

If you pick the right implementation, yes, it is shown twice.

File: hugo/content/docs/recipes/validation/dependency-loops.md (hunk 5 above)
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md:6]: I added some concrete examples

File: hugo/content/docs/recipes/validation/dependency-loops.md (hunk 6 above)
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md]: Extended explanation

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 10]
```diff
@@ -0,0 +1,258 @@
...
+After finishing your validator, do not forget to build your project with `npm run build`.
+So a `.hello` file like this one, would have 3 greetings with an error:
+
+```plaintext
+person Homer
+person Marge
+person Pinky
+person Brain
+
+Homer greets Marge! //error
```
[the remark this answers is not in the document]
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md:211]: Done

File: hugo/content/docs/recipes/validation/dependency-loops.md (hunk 9 above)
[the remark this answers is not in the document]
Comment (JohannesMeierSE) [reply hugo/content/docs/recipes/validation/dependency-loops.md:113]: If you want to discuss only the example, than there is no need to discuss other cases.

**Comment by JohannesMeierSE** [comment]
Thanks @Lotes for the improvements! The two use cases are much more clear in the introduction now and I like the explicit examples.

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+Here is the screenshot of VS Code with the error:
+
+![Greeting loop errors](/assets/dependency-loops.png)
+
+### Dependency resolution
+
+The topological sort can be done like this:
+
+```typescript
+import { topologicalSort } from 'graphology-dag‘;
```
Comment (msujew) [inline hugo/content/docs/recipes/validation/dependency-loops.md:229]: Typo?
```suggestion
import { topologicalSort } from 'graphology-dag';
```

File: hugo/content/docs/recipes/validation/dependency-loops.md [hunk 12]
```diff
@@ -0,0 +1,336 @@
+---
+title: "Dependency loops"
```
Comment (msujew) [inline hugo/content/docs/recipes/validation/dependency-loops.md:2]: Title Case:
```suggestion
title: "Dependency Loops"
```

**Review by msujew — APPROVED** [review]
Looks good, can be merged after resolving the minor issues below.

File: hugo/content/docs/recipes/validation/dependency-loops.md (hunk 11 above)
Comment (Lotes) [reply hugo/content/docs/recipes/validation/dependency-loops.md:229]: Done

**Pull request merged**